text/0000-allow-if-let-irrefutables.md

```diff
@@ -0,0 +1,78 @@
+- Feature Name: allow_if_let_irrefutables
+- Start Date: 2017-07-27
+- RFC PR:
+- Rust Issue:
+
+# Summary
+[summary]: #summary
+
+Currently when using an if let statement and an irrefutable pattern (read always match) is used the compiler complains with an `E0162: irrefutable if-let pattern`.
+The current state breaks macros who want to accept patterns generically and this RFC proposes changing this error to an error-by-default which is allowed to be disabled by such macros.
+
+# Motivation
+[motivation]: #motivation
+
+The use cases for this is in the creation of macros where patterns are allowed because to support the `_` patterns the code has to be rewritten to be both much larger and include a compiler allow.
+The expected outcome is for irrefutable patterns to be compiled to a tautology and have the if block accept it as if it was `if true { }`.
+To support this, currently you must do something roughly the following, which seems to counteract the benefit of having if-let and while-let in the spec.
+
+```rust
+if let $p = $val {
+    $b
+}
+```
+Cannot be used, so the original match must be. The `allow` is forced so that the warning does not appear to the user of it since `_` won't be matched if `$p` is irrefutable.
+
+```rust
+#[allow(unreachable_patterns)]
+match $val {
+    $p => { $b; },
+    _ => ()
+}
+```
+
+
+# Detailed design
+[design]: #detailed-design
+
+1. Change the compiler error `irrefutable if-let-pattern` and similar patterns to an `error-by-default` lint that can be disabled by an `#[allow]` statement
+2. Proposed lint name: `irrefutable_let_pattern`
+
+Code Example (explicit):
+```rust
+#[allow(irrefutable_let_pattern)]
+if let _ = 'a' {
+    println!("Hello World");
+}
+```
+
+Code Example (implicit):
+```rust
+macro_rules! check_five {
+    ($p:pat) => {{
+        #[allow(irrefutable_let_pattern)]
+        if let $p = 5 {
+            println!("Pattern matches five");
+        }
+    }};
+}
+```
+
+# How We Teach This
+[how-we-teach-this]: #how-we-teach-this
+
+This can be taught by changing the second version of (`The Book`)[https://doc.rust-lang.org/book/second-edition/ch18-02-refutability.html] to not explicitly say that it is not allowed.
+Adding that it is a lint that can be disabled.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+It allows programmers to manually write the line `if let _ = expr { } else { }` which is generally obfuscating and not desirable. However, this will not be explicitly allowed with the `#[allow]`.
```

Comment:
You wrote "this will not be explicitly allowed with the `#[allow]`"

Is this a typo? I.e., I would have expected "this will not be explicitly allowed *without* the `#[allow]`"

(Also, I'd probably go ahead and spell out `#[allow(irrefutable_let_pattern]` here)